### ISSUE TITLE
New Plugin - Modified VMs

### ISSUE THREAD

**Issue by omrsafetyo** [alanrenouf/vCheck-vSphere#611]
I have had this mid-2015, and when I pulled the latest copy of vCheck last week, I realized that 2 of the other plugins I had created have been duplicated by someone else, but it looks like this plugin is not.  This Plugin will find any VM with ConfigSpec events, and pull the VMname, start time, operation, device, user, memory (if changed), CPU (if changed) disk information (if changed), and a detail of the device info summary.

I wasn't sure how to label this plugin - I'm not quite sure how the numbering system works.  So let me know if you are interested in including it.

```
$Title = "Modified VMs"
$Header = "VM Modified Events: [count]"
$Display = "Table"
$Author = "Nathan Kulas"
$PluginVersion = 1.0
$PluginCategory = "vSphere"


# Start of Settings 
# Set the number of days to show VMs modified for
$VMsModifiedAge = 2
# End of Settings

$events = @(Get-VIEventPlus -Start ((get-date).adddays(-$VMsModifiedAge)) -EventType "VmReconfiguredEvent")
$report = @()

 ForEach ($event in $events){
	ForEach ( $ConfigSpec in $event.ConfigSpec ) {
		ForEach ( $DeviceChange in $ConfigSpec.DeviceChange ) {
			$report += New-Object PSObject -Property @{
				VMname = $event.vm.Name
				Start = $event.CreatedTime
				Operation = $DeviceChange.Operation
				Device = $DeviceChange.Device.GetType().Name
				User = $event.UserName
				Memory = &{if($ConfigSpec.MemoryMB -ne 0){$ConfigSpec.MemoryMB}else{""}}
				NumCPU = &{if($ConfigSpec.NumCPUs -ne 0){$ConfigSpec.NumCPUs}else{""}}
				Disk = &{if($DeviceChange.Device.GetType().FullName -eq "VMware.Vim.VirtualDisk"){ $DeviceChange.Device.DeviceInfo.Label }else{""}}
				Detail = ($DeviceChange | select -expand Device).DeviceInfo.Summary	
		   }
		}
	}
}

$Comments = "The following VMs modifications were made over the last $($VMsModifiedAge) days"

$report | select VMname, User, Start, Operation, Device, Detail, Memory, NumCPU, Disk

# Change Log
```

**Comment by gregharms** [comment]
You've got my vote for adding it. Getting good output in testing. In my environment, 90% of the events were tripped by backup/replication actions. So added a quick-n-dirty exclusion loop to avoid those events. The exclusion setting is an array of hash tables that can contain arrays themselves. Maybe too complicated but wanted enable multiple types of events to be excluded if necessary. 
Setting:
`# Set events to exclude`
`$eventsToExclude = @(`
`    @{`
`        "User"      = @("DOMAIN\SampleUser1", "DOMAIN\SampleUser2")`
`        "Device"    = "VirtualDisk"`
`        "Operation" = @("add", "remove")`
`    },`
`    @{`
`        "User" = "SampleUserB"`
`        "Device" = "VirtualCdrom"`
`	"Operation" = "edit"`
`    }`
`)`

Loop at the end:
`#Exclude certain events with matching User, Device, and Operation`
`foreach ($exclusion in $eventsToExclude) {`
    `$report = $report | Where-Object {`
        `!(`
            `$_.User -in $exclusion.item("User") -and `
            `$_.Device -in $exclusion.item("Device") -and `
            `$_.Operation -in $exclusion.item("Operation")`
        `)`
    `}`
`}`

**Comment by Sneddo** [comment]
Yeah, great idea for a plugin, but probably needs some filtering options as @gregharms alluded to.

**Comment by omrsafetyo** [comment]
Ah yeah you're right.  This was working well in our legacy vCenters.  We have  a couple vCenters on 5.5, and when we were there, we were using Evault backup software.  I have had vCheck running over there for a couple years.  Recently we started getting some questions about who made what changes, and I realized I had never ported it to any of our new 6.0 vCenters.  So we've had a basically empty report collecting up for the past couple months afterwe retired the last couple 100 Vms out of there.

Since then we have also switched backup vendors, and I'm seeing a lot of VirtualDisk operations for those as well.

Your exclusions work great @gregharms but I'll look at some other plugins that have filtering/exclusions and see if I can find a way so the standard tools already in place can handle the configuration.  Perhaps store the settings as:

`$exclusions` = "User=DOMAIN\backupaccount,Device=VirtualDisk,Operation=add|remove;User=Sample,Device=CdRom,Operation=edit"

So basically, this is your hash table stored as a string.  The fields are comma delimited.  The individual hashtables are colon delimited, and if you want an array to filter on, those are pipe delimited.  So this string basically exactly matches your hashtable.

I can then create a filter block for this as follows:

```
if (-not([string]::IsNullOrEmpty($exclusions))) {
	$FilterArray = @()
	$exclusions -split ";" | ForEach-Object {
		$SubfilterArray = @()
	   $PropArray = $_ -split ","
	   foreach ( $prop in $PropArray ) {
		 $split = $prop -split "="
		 if ( $split[1] -match '|' ) {
			$OrArray = @()
			ForEach ( $value in ($split[1] -split "\|") ) {
				$OrArray += '$_.{0} -eq "{1}"' -f $split[0], $value
			}
			$SubFilterArray += '({0})' -f $($OrArray -join " -or ")
		 }
		 else {
			$SubFilterArray += '$_.{0} -eq "{1}"' -f $split[0], $split[1]
		 }
	   }
	   $FilterArray += "-not({0})" -f $($SubFilterArray -join " -and ")
	}
	$FilterString = $FilterArray -Join " -and "
	$FilterScript = [scriptblock]::Create($FilterString)
	$report = $report | Where-Object -FilterScript $FilterScript 
}
```

This actually just dynamically generates a where block by breaking up the list on the various delimiters, and creating a filter scriptblock, which is passed to the Where-Object command.

This way, I can still prompt for the settings via the -config flag, rather than have people hand edit.

What do you think?

**Comment by omrsafetyo** [comment]
Also, the numbering system appears to be purely arbitrary, is that accurate?

I was going to name this 203 VM Modified, and put it in 60 Vm.  I set a default filter that takes an arbitrary backup user, and excludes add/remove events for vDisks, and then for the CdRom, it excludes all events, regardless of user - because of the way the filter is built, any number of properties can be filtered on at the user's discretion.

`$VmModifiedExclusions = "User=DOMAIN\backupuser,Device=VirtualDisk,Operation=add|remove;Device=VirtualCdRom,Operation=edit"`

**Comment by gregharms** [comment]
Genius. Thank you for introducing me to Where-Object's FilterScript parameter. The dynamic filter worked fine in my testing. I've also added the Unique parameter to the last Select-Object for the report output. For some reason, $events has four events for each change. Adding the Unique parameter was a quick fix.

**Comment by Sneddo** [comment]
> the numbering system appears to be purely arbitrary, is that accurate?

Yeah, pretty much 

Nice solution with the filtering- a good balance between functionality and usability :)
